// verdaccio/build/lib/logger/levels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calculateLevel = calculateLevel;
exports.subsystems = exports.levels = void 0;

var _kleur = require("kleur");

// level to color
const levels = {
  fatal: _kleur.red,
  error: _kleur.red,
  warn: _kleur.yellow,
  http: _kleur.magenta,
  info: _kleur.cyan,
  debug: _kleur.green,
  trace: _kleur.white
};
/**
 * Match the level based on buyan severity scale
 * @param {*} x severity level
 * @return {String} security level
 */

exports.levels = levels;

function calculateLevel(x) {
  switch (true) {
    case x < 15:
      return 'trace';

    case x < 25:
      return 'debug';

    case x < 35:
      return 'info';

    case x == 35:
      return 'http';

    case x < 45:
      return 'warn';

    case x < 55:
      return 'error';

    default:
      return 'fatal';
  }
}

const subsystems = [{
  in: (0, _kleur.green)('<--'),
  out: (0, _kleur.yellow)('-->'),
  fs: (0, _kleur.black)('-=-'),
  default: (0, _kleur.blue)('---')
}, {
  in: '<--',
  out: '-->',
  fs: '-=-',
  default: '---'
}];
exports.subsystems = subsystems;
